File: backend/internal/albumdetector/handler.go
package albumdetector

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"
)

type Handler struct {
	client *Client
}

func NewHandler(client *Client) *Handler {
	return &Handler{client: client}
}

func (h *Handler) Detect(w http.ResponseWriter, r *http.Request) {
	file, header, err := r.FormFile("video")
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		return
	}
	defer file.Close()

	result, err := h.client.Detect(header.Filename, file)
	if err != nil {
		if detErr, ok := err.(*DetectionError); ok {
			w.WriteHeader(detErr.StatusCode)
			w.Header().Set("Content-Type", "application/json")
			w.Write([]byte(detErr.Body))
			return
		}
		w.WriteHeader(http.StatusServiceUnavailable)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(result)
}

func (h *Handler) RegisterRoutes(r chi.Router) {
	r.Post("/detect", h.Detect)
}

File: backend/src/albumdetector/handler.ts
import { Hono } from 'hono'
import { AlbumDetectorClient, DetectionError } from './client.js'

export function albumDetectorRoutes(client = new AlbumDetectorClient()) {
  const routes = new Hono()

  routes.post('/detect', async (c) => {
    const contentType = c.req.header('content-type') ?? ''
    if (!contentType.startsWith('multipart/form-data')) {
      return c.body(null, 400)
    }

    try {
      const body = await c.req.arrayBuffer()
      const upstream = await client.detect(body, contentType)
      return new Response(upstream.body, {
        status: 200,
        headers: { 'Content-Type': upstream.headers.get('content-type') ?? 'application/json' },
      })
    } catch (err) {
      if (err instanceof DetectionError) {
        return c.body(err.body, err.statusCode as 400, { 'Content-Type': 'application/json' })
      }
      return c.body(null, 503)
    }
  })

  return routes
}
